Moving to semantic versioning: 1.0.0 Release (#5217)
skyportal/skyportal: doc/versioning.md

# Versioning

Breaking API changes, and changes that require major upgrades of critical system dependencies (like PostgreSQL), are rare. On the other hand, features (new endpoints, or additional data returned by existing endpoints) are added regularly. Lastly, changes that impact neither the user nor the developer (e.g. refactoring, performance improvements, bug fixes) happen continuously, but do not typically need to be advertised.

SkyPortal uses [Semantic Versioning](https://semver.org/) to indicate the severity of changes. The version number will be of the form `MAJOR.MINOR.PATCH`, where:

- `MAJOR` is incremented when breaking changes, or major dependency upgrades, are introduced;
- `MINOR` is incremented when new features are added in a backwards-compatible manner, requiring no changes to user code and minimal changes to existing deployments; and
- `PATCH` is incremented when backwards-compatible bug fixes are made.

> The development of SkyPortal is active enough that it would be impractical to update the version number for every commit. Instead, we aim for a monthly cadence. New features will be added earlier on in the cycle, then tested, and bugfixes implemented if necessary. We update the version number along with a signed release on GitHub pointing to a specific commit in the repository. That way, the developer / system admin of a SkyPortal instance can easily checkout to the latest tagged release's commit rather than directly point to the main branch's last commit. The end-user can easily check which version of SkyPortal they are querying, as it is returned in every API response. New versions will also be announced using the **Discussions** page of the repository. The latest available version number is also visible at the top left of the [Documentation](https://skyportal.io/docs) page.

Reviewer: ```suggestion
We aim to make releases monthly. New features are added earlier in the release cycle, then tested, with bugfixes made as necessary. New versions are published on GitHub, generated from tagged commits in the repository. System administrators would typically deploy directly from the tag. End-users can see, from within SkyPortal, which version they are querying: this information is included in API responses, as well as on the About page.
```

I think we should stop announcing on discussions, and rely on GitHub releases to be the One Place where users should look.

# Versioning

SkyPortal being an API-based service, its version number should have a meaning for the end-user AND the developer or system administrator that will deploy it. On the other hand, maintaining 2 distinct version numbers would not be practical.

> The development cycle of SkyPortal is such that breaking API changes are rare, as we often add, improve, refactor, or fix existing features rather than change their output or behavior. Version changes of critical system dependencies are also very rare (e.g. upgrading the version of PostgreSQL or the Python runtime). On the other hand, new features (new endpoints, new tables, additional data returned by existing endpoints) are added more frequently. Lastly, changes that do not impact the user nor the developer (e.g. refactoring, performance improvements, bug fixes) are very common, but do not necessarily need to be advertised to the end-user.

Reviewer: ```suggestion
Breaking API changes, and changes that require major upgrades of critical system dependencies (like PostgreSQL), are rare. On the other hand, features (new endpoints, or additional data returned by existing endpoints) are added regularly. Lastly, changes that impact neither the user nor the developer (e.g. refactoring, performance improvements, bug fixes) happen continuously, but do not typically need to be advertised.
```

> Based on these considerations, we have decided to use [Semantic Versioning](https://semver.org/) for SkyPortal. The version number will be of the form `MAJOR.MINOR.PATCH`, where:

Reviewer: ```suggestion
SkyPortal uses [Semantic Versioning](https://semver.org/) to indicate the severity of changes. The version number will be of the form `MAJOR.MINOR.PATCH`, where:
```

> - `MAJOR` is incremented when breaking changes are introduced, for the user or the developer

Reviewer: ```suggestion
- `MAJOR` is incremented when breaking changes, or major dependency upgrades, are introduced;
```

> - `MINOR` is incremented when new features are added in a backwards-compatible manner, requiring no changes to the user's code and minimal changes to the deployment

Reviewer: ```suggestion
- `MINOR` is incremented when new features are added in a backwards-compatible manner, requiring no changes to user code and minimal changes to existing deployments; and
```

> - `PATCH` is incremented when backwards-compatible bug fixes are introduced, or when changes are made that do not impact the user nor the developer

Reviewer: ```suggestion
- `PATCH` is incremented when backwards-compatible bug fixes are made.
```

> The development of SkyPortal is active enough that it would be impractical to update the version number for every commit. Instead, we will update the version number along with a release on GitHub that tags a specific commit in the repository, at most one a week. That way, the developer / system admin can easily checkout to the latest tagged release's commit (when there is a new version available) rather than simply point to whatever is on main. The end-user can always easily check which version of SkyPortal they are querying/using, as it is returned in every API response and displayed on the about page of the web application.

Reviewer: I reviewed up to here; can I capture what was discussed on the call this morning for this sentence? I.e., we have, say, ~monthly release cadences, where new features get introduced early in the cycle, then tested, and a release is made once everything's stabilized?

Author: Yep! I need to edit that based on the call indeed

Author: Edited, please let me know what you think!

Each instance of SkyPortal is responsible for updating the version of the software they are running, but it is good to keep in mind that the lastest version of SkyPortal is always available on GitHub, and that the latest version number is also visible at the top left of the [Documentation](https://skyportal.io/docs) page.

> The version number will be updated in the `skyportal/__init__.py` file, and release notes will be posted with the tagged release.

Reviewer: ```suggestion
The SkyPortal version number is accessible as `skyportal.__version__`.
```

> As a SkyPortal, please refer to the [Developer Guide](dev) for more information on how to interact with the codebase, and to the [Deployment Guide](deploy) for more information on how to deploy SkyPortal.

Reviewer: ```suggestion
Please refer to the [Deployment Guide](deploy) for more information on how to deploy SkyPortal.
```
